File: pkg/store/sync_state_store.go
package store

import (
	"context"
	"errors"
	"fmt"
	"strconv"

	"gorm.io/gorm"
	"gorm.io/gorm/clause"

	"github.com/PineappleBond/xyncra-server/pkg/store/model"
)

// SyncStateStore provides key-value data access for client-side synchronization
// state tracking (e.g. local_max_seq, latest_seq).
type SyncStateStore struct {
	db *gorm.DB
}

const (
	// syncKeyLocalMaxSeq is the key for the local maximum processed seq.
	syncKeyLocalMaxSeq = "local_max_seq"
	// syncKeyLatestSeq is the key for the server-reported latest seq.
	syncKeyLatestSeq = "latest_seq"
)

// Get retrieves the value for the given key. Returns ErrNotFound if the key
// does not exist.
func (ss *SyncStateStore) Get(ctx context.Context, key string) (string, error) {
	var state model.SyncState
	err := ss.db.WithContext(ctx).
		Where("key = ?", key).
		First(&state).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return "", ErrNotFound
		}
		return "", classifyError(fmt.Errorf("store: get sync state: %w", err))
	}
	return state.Value, nil
}

// Set performs an UPSERT for the given key-value pair. If the key already
// exists, the value is updated; otherwise a new record is inserted.
func (ss *SyncStateStore) Set(ctx context.Context, key, value string) error {
	state := model.SyncState{Key: key, Value: value}
	err := ss.db.WithContext(ctx).
		Clauses(clause.OnConflict{
			Columns:   []clause.Column{{Name: "key"}},
			DoUpdates: clause.AssignmentColumns([]string{"value", "updated_at"}),
		}).
		Create(&state).Error
	if err != nil {
		return classifyError(fmt.Errorf("store: set sync state: %w", err))
	}
	return nil
}

// GetLocalMaxSeq returns the local_max_seq value. Returns 0 if not set.
func (ss *SyncStateStore) GetLocalMaxSeq(ctx context.Context) (uint32, error) {
	val, err := ss.Get(ctx, syncKeyLocalMaxSeq)
	if err != nil {
		if errors.Is(err, ErrNotFound) {
			return 0, nil
		}
		return 0, err
	}
	seq, err := strconv.ParseUint(val, 10, 32)
	if err != nil {
		return 0, fmt.Errorf("store: parse local_max_seq: %w", err)
	}
	return uint32(seq), nil
}

// SetLocalMaxSeq sets the local_max_seq value.
func (ss *SyncStateStore) SetLocalMaxSeq(ctx context.Context, seq uint32) error {
	return ss.Set(ctx, syncKeyLocalMaxSeq, strconv.FormatUint(uint64(seq), 10))
}

// GetLatestSeq returns the latest_seq value. Returns 0 if not set.
func (ss *SyncStateStore) GetLatestSeq(ctx context.Context) (uint32, error) {
	val, err := ss.Get(ctx, syncKeyLatestSeq)
	if err != nil {
		if errors.Is(err, ErrNotFound) {
			return 0, nil
		}
		return 0, err
	}
	seq, err := strconv.ParseUint(val, 10, 32)
	if err != nil {
		return 0, fmt.Errorf("store: parse latest_seq: %w", err)
	}
	return uint32(seq), nil
}

// SetLatestSeq sets the latest_seq value.
func (ss *SyncStateStore) SetLatestSeq(ctx context.Context, seq uint32) error {
	return ss.Set(ctx, syncKeyLatestSeq, strconv.FormatUint(uint64(seq), 10))
}

File: demo/vue-pure-admin/packages/xyncra-client-core/src/db/sync-state-store.ts
/**
 * SyncStateStore — key-value data access for client-side synchronization state.
 *
 * Mirrors Go SyncStateStore (pkg/store/sync_state_store.go).
 *
 * Well-known keys:
 *   - local_max_seq: the local maximum processed seq.
 *   - latest_seq: the server-reported latest seq.
 */

import type Dexie from 'dexie';
import type { Transaction } from 'dexie';

import { ErrNotFound } from '../errors';
import type { XyncraDatabase } from './index';
import type { SyncState } from './models';
import { SyncStateKey } from './models';

/**
 * SyncStateStore provides key-value data access for client-side synchronization
 * state tracking (e.g. local_max_seq, latest_seq).
 */
export class SyncStateStore {
  constructor(private readonly db: XyncraDatabase) {}

  /**
   * Retrieves the value for the given key.
   * Throws ErrNotFound if the key does not exist.
   */
  async get(key: string): Promise<string> {
    const state = await this.db.syncStates.get(key);
    if (!state) {
      throw ErrNotFound;
    }
    return state.value;
  }

  /**
   * Performs an UPSERT for the given key-value pair.
   * If the key already exists, the value is updated; otherwise a new record is
   * inserted.
   */
  async set(key: string, value: string): Promise<void> {
    const now = new Date();
    await this.db.syncStates.put({
      key,
      value,
      updated_at: now,
    });
  }

  /**
   * Returns the local_max_seq value. Returns 0 if not set.
   */
  async getLocalMaxSeq(): Promise<number> {
    try {
      const val = await this.get(SyncStateKey.LocalMaxSeq);
      const seq = Number.parseInt(val, 10);
      return Number.isNaN(seq) ? 0 : seq;
    } catch (error) {
      if (error === ErrNotFound) return 0;
      throw error;
    }
  }

  /**
   * Sets the local_max_seq value.
   */
  async setLocalMaxSeq(seq: number): Promise<void> {
    await this.set(SyncStateKey.LocalMaxSeq, String(seq));
  }

  /**
   * Returns the latest_seq value. Returns 0 if not set.
   */
  async getLatestSeq(): Promise<number> {
    try {
      const val = await this.get(SyncStateKey.LatestSeq);
      const seq = Number.parseInt(val, 10);
      return Number.isNaN(seq) ? 0 : seq;
    } catch (error) {
      if (error === ErrNotFound) return 0;
      throw error;
    }
  }

  /**
   * Sets the latest_seq value.
   */
  async setLatestSeq(seq: number): Promise<void> {
    await this.set(SyncStateKey.LatestSeq, String(seq));
  }

  /**
   * Sets local_max_seq within the given transaction.
   */
  async setLocalMaxSeqTx(tx: Transaction, seq: number): Promise<void> {
    const table = tx.table('syncStates') as Dexie.Table<SyncState, string>;
    await table.put({
      key: SyncStateKey.LocalMaxSeq,
      value: String(seq),
      updated_at: new Date(),
    });
  }
}
